processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -730,6 +763,45 @@ public DateTime getMaxIngestedEventTime()
     return maxIngestedEventTime;
   }
 
+  private class DimensionHolder
+  {
+    @GuardedBy("lock")
+    private DimDim[] dimensions;
+    private final Object lock;
+
+    private DimensionHolder(int numDimension, Object lock)
+    {
+      this.dimensions = new DimDim[Math.max(32, numDimension)];
```

Comment:
I'm also in favor of changing the arrays that require explicit resizing (DimensionHolder, OnHeapDimDim) to lists